### PR TITLE
Refactor KAFKA processor to wherehows-kafka module

### PR DESCRIPTION
**Pull request by hangzhang925**
Some key points for this PR:
**dependency:**
Added slf4j_simple for lombok slf4j implementation (for wherehows-kafka)

**wherehows-backend:**
Backend will keep the same, we can remove the kafka part once we launch MP wherehows-kafka.

**wherehows-dao**
1. Add AbstractDao for some common operation like update, delete ...
2. Add new models and related DAOs
	new file:   ../wherehows-dao/src/main/java/wherehows/models/CFGApplication.java
	new file:   ../wherehows-dao/src/main/java/wherehows/models/ClusterInfo.java
	new file:   ../wherehows-dao/src/main/java/wherehows/models/DatasetFieldSchema.java
	new file:   ../wherehows-dao/src/main/java/wherehows/models/DatasetSchemaInfo.java
	new file:   ../wherehows-dao/src/main/java/wherehows/models/DictDataset.java
	new file:   ../wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java

**wherehows-kafka**
1. refactor the code using DAO-SERVICE-CONTROLLER pattern.
Old DAO contains data access and logical part of the application.
New DAO:  only contains data access, SERVICE responsible for all application logical.
Old process:   [kAFKA MESSAGE] -> processor -> DAO(save to different DB, update data format....)
New process:   [kAFKA MESSAGE] -> processor -> SERVICE (save to different DB, update data format....) -> DAO (CRUD)

2.  use gradle application plugin
Main class will initialize DAOFactory programmatically, check jobs folder and start service.

3. removed DB writer
processor.record is void function now, processorService layer will complete logical part and call DAO to save data.

**TEST DONE**
local build passed for all modules.
Able to run wherehows-kafka on stg server and receive/process KAFKA event. 

**TODO**
write test case for wherehow-kafka module

### CODE REVIEW

File: wherehows-dao/src/main/java/wherehows/dao/AbstractDao.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package wherehows.dao;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+
+
+public class AbstractDao {
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/AbstractDao.java]: Rename this to BaesDao since there's nothing abstract about it?

File: wherehows-dao/src/main/java/wherehows/dao/DaoFactory.java [hunk 2]
```diff
@@ -35,4 +35,20 @@ public DaoFactory(EntityManagerFactory entityManagerFactory) {
   public DatasetClassificationDao getDatasetClassificationDao() {
     return new DatasetClassificationDao(entityManagerFactory);
   }
+
+  public DictDatasetDao getDictDatasetDao() {
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/DaoFactory.java]: Let's drop the Dict prefix

File: wherehows-kafka/src/main/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+hikaricp.dataSourceClassName = "com.mysql.jdbc.jdbc2.optional.MysqlDataSource"
+hikaricp.dataSourceClassName = ${?WHZ_DB_DSCLASSNAME}
+
+# dialect for DB
+# example values for different db: org.hibernate.dialect.MySQL5InnoDBDialect, org.hibernate.dialect.H2Dialect
+hikaricp.dialect = "org.hibernate.dialect.MySQL5InnoDBDialect"
+hikaricp.dialect = ${?WHZ_DB_DIALECT}
+
+# example values for different db: org.hibernate.dialect.MySQL5InnoDBDialect, org.hibernate.dialect.H2Dialect
+etl.jobs.dir = ${WHZ_ETL_JOBS_DIR}
```
Comment (mars-lan) [inline wherehows-kafka/src/main/resources/application.conf]: Are these still needed?

File: settings.gradle [hunk 4]
```diff
@@ -7,6 +7,7 @@ def modules = [
     "wherehows-frontend",
     "wherehows-hadoop",
     "wherehows-web",
+    "wherehows-kafka",
```
Comment (mars-lan) [inline settings.gradle]: Order alphabetically

File: wherehows-dao/src/main/java/wherehows/dao/CFGApplicationDao.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/**
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/CFGApplicationDao.java]: Is this file used anywhere?

File: wherehows-dao/src/main/java/wherehows/dao/DatasetOwnerDao.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/**
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetOwnerDao.java]: Is this used anywhere?

File: wherehows-dao/src/main/java/wherehows/dao/CFGApplicationDao.java (hunk 5 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/CFGApplicationDao.java]: Can i leave the file here? I created it when refactor MetadataChangeProcessor, later on we discussed leave the processor empty and do refactor work in the future.
I believe this will be needed in the future.

File: wherehows-dao/src/main/java/wherehows/dao/DaoFactory.java (hunk 2 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/DaoFactory.java]: Dataset.java(same thing as DictDatasetClass) class right now is used in frontend and backend in a lot of places. Change DictDataset.java to Dataset.java means i need to merge those two classes, how about i do the refactoring work in another PR?

File: wherehows-kafka/src/main/resources/application.conf (hunk 3 above)
Comment (hangzhang925) [inline wherehows-kafka/src/main/resources/application.conf]: nope, i will remove it

File: wherehows-dao/src/main/java/wherehows/dao/AbstractDao.java (hunk 1 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/AbstractDao.java]: good idea. Will rename it

File: settings.gradle [hunk 7]
```diff
@@ -6,7 +6,9 @@ def modules = [
     "wherehows-etl",
     "wherehows-frontend",
     "wherehows-hadoop",
+    "wherehows-kafka",
     "wherehows-web",
+
```
Comment (mars-lan) [inline settings.gradle]: Please also remove the extra line

File: wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+  public DatasetSchemaInfoDao(EntityManagerFactory factory) {
+    super(factory);
+  }
+
+  @SneakyThrows
+  public DatasetSchemaInfo findById(int datasetId) {
+    EntityManager entityManager = entityManagerFactory.createEntityManager();
+    try {
+      return entityManager.createQuery(FIND_BY_DATASET_ID, DatasetSchemaInfo.class)
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java]: Maybe use Criteria instead? https://docs.jboss.org/hibernate/orm/3.5/api/org/hibernate/Criteria.html

File: wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  @Column(name = "jdbc_data_type")
+  String jdbcDataType;
+
+  @Column(name = "pig_data_type")
+  String pigDataType;
+
+  @Column(name = "hcatalog_data_type")
+  String hcatalogDataType;
+
+  Date modified;
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java]: Let's annotate every single field to make it super clear?

File: wherehows-kafka/application.env [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+WHZ_DB_URL="jdbc:mysql://localhost/wherehows"
+
+#mysql setup
+WHZ_DB_DSCLASSNAME = "com.mysql.jdbc.jdbc2.optional.MysqlDataSource"
+WHZ_DB_DIALECT = "org.hibernate.dialect.MySQLInnoDBDialect"
+
+# Directory containing ETL job files
+WHZ_ETL_JOBS_DIR="/var/tmp/jobs"
+
+# Temp directory for ETL job
```
Comment (mars-lan) [inline wherehows-kafka/application.env]: Don't think this is needed?

File: wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java (hunk 8 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java]: Sure, i think the criteria way is better we don't need to hard code table name/queries in our code

File: wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java (hunk 9 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java]: Added.

File: settings.gradle (hunk 7 above)
Comment (hangzhang925) [inline settings.gradle]: removed

File: wherehows-kafka/application.env (hunk 10 above)
Comment (hangzhang925) [inline wherehows-kafka/application.env]: updated

File: wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+  public DatasetSchemaInfoDao(EntityManagerFactory factory) {
+    super(factory);
+  }
+
+  @SneakyThrows
+  public DatasetSchemaInfo findById(int datasetId) {
+    EntityManager entityManager = entityManagerFactory.createEntityManager();
+    CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+    CriteriaQuery<DatasetSchemaInfo> criteria = cb.createQuery(DatasetSchemaInfo.class);
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java]: You can probably do some chaining here to make it more readable, e.g. cretiera.select().where()

File: wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    Root<DatasetSchemaInfo> entityRoot = criteria.from(DatasetSchemaInfo.class);
+    criteria.select(entityRoot);
+    criteria.where(cb.equal(entityRoot.get("dataset_id"), datasetId));
+
+    try {
+      return entityManager.createQuery(criteria)
```
Comment (mars-lan) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java]: Why not juse use https://docs.jboss.org/hibernate/orm/3.5/api/org/hibernate/Criteria.html#uniqueResult()?

Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java]: Thanks.
Actually uniqueResult is for session.createCritria, but i found a better solution for my case.  entityManager.createQuery(criteria).getSingleResult();
https://docs.jboss.org/hibernate/entitymanager/3.5/reference/en/html/querycriteria.html#querycriteria-typedquery-expression

File: wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java (hunk 11 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/DatasetSchemaInfoDao.java]: done

File: wherehows-kafka/src/main/java/wherehows/utils/JobsUtil.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
+/**
```
Comment (mars-lan) [inline wherehows-kafka/src/main/java/wherehows/utils/JobsUtil.java]: Maybe move this to wherehows-common in the next PR?

File: wherehows-dao/src/main/java/wherehows/dao/ClusterInfoDao.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+
+public class ClusterInfoDao extends BaseDao {
+
+  public ClusterInfoDao(EntityManagerFactory factory) {
+    super(factory);
+  }
+
+  @SneakyThrows
+  public List<ClusterInfo> findAll() {
```
Comment (alyiwang) [inline wherehows-dao/src/main/java/wherehows/dao/ClusterInfoDao.java]: Later on, can maybe move this to BaseDao also

File: wherehows-kafka/src/main/java/wherehows/utils/JobsUtil.java (hunk 13 above)
Comment (hangzhang925) [inline wherehows-kafka/src/main/java/wherehows/utils/JobsUtil.java]: My understanding is we are trying to not rely on wherehows-common. 
Once we refactored all models to wherehows-dao, we can remove the dependency :)
But let me know if you have any concerns.

File: wherehows-dao/src/main/java/wherehows/dao/ClusterInfoDao.java (hunk 14 above)
Comment (hangzhang925) [inline wherehows-dao/src/main/java/wherehows/dao/ClusterInfoDao.java]: OK, sounds good!

File: wherehows-kafka/build.gradle [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    compile externalDependency.guava
+    compile externalDependency.slf4j_simple
+}
+
+configurations {
+    all*.exclude group: 'org.slf4j', module: 'slf4j-log4j12'
+    all*.exclude group: 'log4j'
+
+    all*.resolutionStrategy {
+        dependencySubstitution {
```
Comment (alyiwang) [inline wherehows-kafka/build.gradle]: Those substitution may not be needed here, since the dependency is relatively clean. 
Those are mainly for backend which included ETL and play

File: wherehows-kafka/src/main/java/wherehows/main/ApplicationStart.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  public static void main(String[] args) {
+    ClusterInfoDao clusterInfoDao = DAO_FACTORY.getClusterInfoDao();
+    ClusterUtil.updateClusterInfo(clusterInfoDao.findAll());
+
+    ActorSystem actorSystem = ActorSystem.create("WhereHowsETLService");
+    ActorRef kafkaMaster = actorSystem.actorOf(Props.create(KafkaConsumerMaster.class), "KafkaMaster");
+
+
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/main/ApplicationStart.java]: remove an empty line

File: wherehows-kafka/src/main/java/wherehows/processors/GobblinTrackingAuditProcessor.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    // only handle "DaliLimitedRetentionAuditor","DaliAutoPurgeAuditor" and "DsIgnoreIDPCAuditor"
+    if (name.equals(DALI_LIMITED_RETENTION_AUDITOR) || name.equals(DALI_AUTOPURGED_AUDITOR) || name.equals(
+        DS_IGNORE_IDPC_AUDITOR)) {
+      // TODO: Re-enable this once it's fixed.
+    } else if (name.equals(METADATA_FILE_CLASSIFIER)) {
+      gobblinTrackingAuditService.updateHdfsDatasetSchema(record);
+    }
+  }
+
+
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/GobblinTrackingAuditProcessor.java]: too many empty lines

File: wherehows-kafka/build.gradle (hunk 15 above)
Comment (hangzhang925) [inline wherehows-kafka/build.gradle]: OK, removed the resolutionStrategy

File: wherehows-kafka/src/main/java/wherehows/main/ApplicationStart.java (hunk 16 above)
Comment (hangzhang925) [inline wherehows-kafka/src/main/java/wherehows/main/ApplicationStart.java]: removed

File: wherehows-kafka/src/main/java/wherehows/processors/GobblinTrackingAuditProcessor.java (hunk 17 above)
Comment (hangzhang925) [inline wherehows-kafka/src/main/java/wherehows/processors/GobblinTrackingAuditProcessor.java]: removed

File: wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+import javax.persistence.Id;
+import javax.persistence.Table;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+
+@Data
+@Entity
+@Table(name = "dataset_classification")
```
Comment (alyiwang) [inline wherehows-dao/src/main/java/wherehows/models/DictFieldDetail.java]: table name is incorrect
